File: packages/typescript/test/contract.test.ts
// 契约测试：以根 tests/ 的 Schema + Fixture 验证 TypeScript 模型。
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { dirname, join } from 'node:path';
import { describe, expect, it } from 'vitest';
import { parseComment, parseSolution, parseTicket } from '../src/models';

const here = dirname(fileURLToPath(import.meta.url));
const fixturesDir = join(here, '../../../tests/fixtures');
const schemasDir = join(here, '../../../tests/schemas');

function fixture(name: string): unknown {
  return JSON.parse(readFileSync(join(fixturesDir, `${name}.json`), 'utf8'));
}

function hasRequired(data: Record<string, unknown>, schemaName: string) {
  const schema = JSON.parse(
    readFileSync(join(schemasDir, `${schemaName}.json`), 'utf8'),
  ) as { required?: string[] };
  const missing = (schema.required ?? []).filter((k) => !(k in data));
  expect(missing).toEqual([]);
}

describe('Ticket 契约', () => {
  it('fixture 覆盖必填字段', () => {
    const data = fixture('ticket') as Record<string, unknown>;
    hasRequired(data, 'ticket');
    const comments = data['comments'] as unknown as Record<string, unknown>[];
    hasRequired(comments[0]!, 'comment');
  });

  it('解析且关键字段符合预期', () => {
    const t = parseTicket(fixture('ticket'));
    expect(t.id).toBe('t-0192');
    expect(t.number).toBe(192);
    expect(t.state).toBe('closed');
    expect(t.author).toBe('黄亮');
    expect(t.labels).toEqual(['faq']);
    expect(t.comments).toHaveLength(2);
    // 人工/AI 区分是 comment.author 的属性
    expect(t.comments[1]?.author).toBe('ai');
    expect(t.closed_at).toBe('2026-09-01T17:16:00+08:00');
  });

  it('round-trip：序列化回 JSON 再解析值不变', () => {
    const t = parseTicket(fixture('ticket'));
    const again = parseTicket(JSON.parse(JSON.stringify(t)));
    expect(again).toEqual(t);
  });

  it('非法 state 拒绝', () => {
    const bad = { ...(fixture('ticket') as object), state: 'done' };
    expect(() => parseTicket(bad)).toThrow(/state/);
  });

  it('open 态无 closed_at', () => {
    const t = parseTicket(fixture('ticket'));
    t.state = 'open';
    delete t.closed_at;
    expect('closed_at' in t).toBe(false);
  });
});

describe('Comment 契约', () => {
  it('解析', () => {
    const data = fixture('ticket') as { comments: unknown[] };
    const c = parseComment(data.comments[0]);
    expect(c.author).toBe('赵子奕');
    expect(c.created_at).toBe('2026-09-01T17:14:00+08:00');
  });
});

describe('Solution 契约', () => {
  it('fixture 覆盖必填字段 + 解析', () => {
    const data = fixture('solution') as Record<string, unknown>;
    hasRequired(data, 'solution');
    const s = parseSolution(data);
    expect(s.id).toBe('s-voucher-usage');
    expect(s.title).toBe('代金券怎么用、怎么挣');
    expect(s.content).toContain('来源：Ticket #192');
  });

  it('round-trip', () => {
    const s = parseSolution(fixture('solution'));
    const again = parseSolution(JSON.parse(JSON.stringify(s)));
    expect(again).toEqual(s);
  });
});

File: packages/dart/test/contract_test.dart
// 契约测试：以根 tests/ 的 Schema + Fixture 验证 Dart 模型。
// 数据形状与 docs/specification 的 JSON 示例对齐（snake_case）。
import 'dart:convert';
import 'dart:io';

import 'package:test/test.dart';
import 'package:quanttide_support/quanttide_support.dart';

final fixturesDir = Directory('${Directory.current.path}/../../tests/fixtures');
final schemasDir = Directory('${Directory.current.path}/../../tests/schemas');

Map<String, dynamic> fixture(String name) =>
    jsonDecode(File('${fixturesDir.path}/$name.json').readAsStringSync())
        as Map<String, dynamic>;

void hasRequired(Map<String, dynamic> data, String schemaName) {
  final schema = jsonDecode(
    File('${schemasDir.path}/$schemaName.json').readAsStringSync(),
  ) as Map<String, dynamic>;
  final missing =
      ((schema['required'] as List?) ?? []).cast<String>().where((k) => !data.containsKey(k));
  expect(missing, isEmpty, reason: '$schemaName 缺少必填字段');
}

void main() {
  test('ticket 契约 + 字段语义', () {
    final data = fixture('ticket');
    hasRequired(data, 'ticket');
    final t = Ticket.fromJson(data);
    expect(t.id, 't-0192');
    expect(t.number, 192);
    expect(t.state, 'closed');
    expect(t.author, '黄亮');
    expect(t.labels, ['faq']);
    expect(t.comments.length, 2);
    // 人工/AI 区分是 comment.author 的属性，不单列字段
    expect(t.comments[1].author, 'ai');
    expect(t.closedAt, '2026-09-01T17:16:00+08:00');
  });

  test('ticket round-trip', () {
    final data = fixture('ticket');
    final t = Ticket.fromJson(data);
    final again = Ticket.fromJson(t.toJson());
    expect(again.id, t.id);
    expect(again.number, t.number);
    expect(again.comments.length, t.comments.length);
    expect(again.comments[1].body, t.comments[1].body);
    expect(again.closedAt, t.closedAt);
  });

  test('ticket open 态无 closed_at', () {
    final t = Ticket(
      id: 't-9999',
      number: 9999,
      title: '测试工单',
      body: '正文',
      state: 'open',
      author: '客户甲',
      createdAt: '2026-09-01T10:00:00+08:00',
    );
    expect(t.toJson().containsKey('closed_at'), isFalse);
  });

  test('comment 契约', () {
    final data = fixture('ticket')['comments'] as List<dynamic>;
    hasRequired(data.first as Map<String, dynamic>, 'comment');
    final c = Comment.fromJson(data.first as Map<String, dynamic>);
    expect(c.author, '赵子奕');
    expect(c.createdAt, '2026-09-01T17:14:00+08:00');
  });

  test('solution 契约 + round-trip', () {
    final data = fixture('solution');
    hasRequired(data, 'solution');
    final s = Solution.fromJson(data);
    expect(s.id, 's-voucher-usage');
    expect(s.title, '代金券怎么用、怎么挣');
    expect(s.content, contains('来源：Ticket #192'));

    final again = Solution.fromJson(s.toJson());
    expect(again.id, s.id);
    expect(again.content, s.content);
  });
}
